### ISSUE TITLE
Error starting Joystick Gremlin - Solution enclosed

### ISSUE THREAD

**Issue by NomDeMorte** [WhiteMagic/JoystickGremlin#224]
TL;DR Joystick Gremlin needs a working path specified, in order to start (Guessing it's because of this https://github.com/WhiteMagic/JoystickGremlin/blob/develop/joystick_gremlin.py#L39 ) Accordingly, it cannot use the 'run' registry keys. A Scheduled Task, can work.

Also, don't change the installation path when using the MSI installer.

Pardon me if this should be added to an existing issue, I've seen several similar but some are closed and some aren't quite the same.

Also, I think I'm actually reporting at least two issues here so perhaps this should be broken into separate parts. Let me know what would be most appropriate.

I'll start at the beginning.

Part 1) Installer issues

Part 1) a) Installation location

I used the MSI installer. I changed the installation path to C:\Program Files(x86)\Joystick Gremlin (just removed the brand name so it's easier for me to find). 

All seemed to work OK, but once I exited the application, when I used the desktop icon to start it again, it would request admin rights to start the installer before launching the app. After this, the app would work as expected.

I removed the desktop icon and used the start menu Icon. The behaviour was the same.

I tried to just run the .exe, and the application appeared to function correctly, visually, but no output ever reached vJoy. 

Upon reading some bug reports here, I uninstalled, and reinstalled to the default directory. This seemed to resolve the issue of requesting admin and reinstalling every time I started the application.

Part 1) a) Workaround: Install to the default location, and use the shortcut icons to start the application.

Part 1) b) Startup entry

I rebooted, and the application did not start automatically as I had set in the options. Task manager showed an entry to start an application, but the entry was blank. I opened the JG options dialog, unchecked and rechecked the box. The entry was recreated in the startup tab in task manager.

Part 1) b) Workaround: 
For uninstallation: Manually delete the startup entry from the registry key Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Run
For re-installation: Untick and re-tick the checkbox in the Options dialog -  except, don't, because it doesn't work. See part 2)

Part 2) Automatic start error

I rebooted, and JG failed to start again. I received the following error message: 
Failed to execute script joystick_gremlin
![image](https://user-images.githubusercontent.com/45681208/61571144-c1a33c80-aad4-11e9-8140-2c5ec905cb02.png)
I've seen various reports of this.

Nothing at all is logged in either debug.txt, system.log or user.log. If I delete these files, they are not recreated.

I tried removing the startup (run key) entry, instead starting the application from a scheduled task, including with a delay of up to 1 minute to ensure the system was fully initialised, and the behaviour was the same.

Starting the application immediately following the error or even prior to it results in normal execution, so I'm sure that everything is functioning normally. Obviously this is some issue with the way that the application is launched.

I realised that the script referenced in the error, was the python script which was the source of the application, so I knew the application was attempting to start, but failing prior to beginning the logging functions. I read through the source and the main function included references to setting the working directory, based on the directory from which the application was launched. I suspected this may be failing.

I tried to set a working directory in the Scheduled Task, et voila. Works like a charm.

TL;DR at the top contains a link to what I assume to be the offending code. 

Scheduled Task is as follows: 

![image](https://user-images.githubusercontent.com/45681208/61572892-1dc38c00-aae8-11e9-8e05-450bb62dc09f.png)

![image](https://user-images.githubusercontent.com/45681208/61572894-2320d680-aae8-11e9-9810-e4ec775dd97b.png)

![image](https://user-images.githubusercontent.com/45681208/61572895-27e58a80-aae8-11e9-8a16-d2d730eed755.png)

![image](https://user-images.githubusercontent.com/45681208/61572900-3764d380-aae8-11e9-9ed3-c50d8391bb1e.png)

![image](https://user-images.githubusercontent.com/45681208/61572902-3b90f100-aae8-11e9-960a-2178c9830f4e.png)

Hope that helps.

**Comment by WhiteMagic** [comment]
Thanks for the detailed comments. I think those are two issues as you pointed out, one the MSI installer being finicky which I haven't touched in ages since I got the first one working. The other one is launch issues when Gremlin may be launched via some "forwarding" function.

I'm not sure if the particular code line is at fault, it might be if the executable is not Gremlin itself, however, the Fatal Error happens, as far as I can tell at least, before any of Gremlin's code is executed, as in it's somewhere within the PyInstaller generated wrapper stuff that fails for unknown reason and no Gremlin code gets run. Though that is just an assumption of mine as I never tracked down how that error is caused.

**Comment by NomDeMorte** [comment]
Would you like me to split off that top part regarding the MSI? I think that's been somewhat covered elsewhere, perhaps I should just remove it?


For an easy way to replicate the fault, just create a shortcut to the exe (rightclick, create shortcut), edit the properties, and set the working directory ("Start in:") to anything other than the path to the exe.

This is why I figured it was that line that set the working directory, failing to do so. The only way it works, is if you set it manually when you launch it. If that line were working, then it would certainly have to be something above line 39 (after which the CWD would be set to the functional path), which is entirely consistent of imports.

It's entirely possible it's some kind of issue with the packaging to an exe though, and it's never even getting that far. There are some relevant notes either way [over here](https://pythonhosted.org/PyInstaller/runtime-information.html)

I guess it would be easy for you to pin this down by simply executing the .py file with the incorrect working directory set. I'd have done this myself but I wasn't sure about the prerequisites required to run it. If you'd like, I could do that for you?

Just would like to give something back to say thanks for a great app :)

**Comment by WhiteMagic** [comment]
No need to create a separate issue for that. And yeah playing around with the working directory is rather easy and probably one bit of the puzzle, I'll see when I look into it.

**Comment by NomDeMorte** [comment]
Looking into this further, I'm fairly confident that you are right and this is an issue with the PyInstaller behaviour. I'm not confident enough in my ability to know the exact issue or solution, but I've found some details which isolate it fairly specifically.

I had a reply written up but it's become exceedingly long. Instead I think I will suggest to look at what I did - I used SysInternals Process Monitor, and looked at the file access. To direct you to the area of interest, find the 5th "Thread Create" operation. Until this point, the behaviour between either success or failure modes, is the same. After this, we can see attempted accesses of temporary directories which do not exist such as C:\Temp, and on my PC I see _your_ user directory temp path being accessed. It's interesting that this is also effecting dill and its logging location. The dll was being touched, in the correct location, prior to this, as were Qt libs, but after this point it no longer uses the correct path. For reference, the debug.txt file is created by dill, but it is never written to.

Here's a screenshot to hopefully save you some time: Success is on the left, failure on the right
![image](https://user-images.githubusercontent.com/45681208/61584193-dcce8480-ab86-11e9-8ffc-f6261c11fb96.png)

Hope I'm being helpful and not just noisey here. Please let me know if there's anything I can do to be of assistance.

**Comment by WhiteMagic** [comment]
Definitely not noise :-) I'm grateful for you looking into this as the issues seem somewhat connected / related to the whole build process. I'll have to look into this at some point in the future, as I'm currently breaking the entire code base to implement a new feature.

**Comment by WhiteMagic** [comment]
Closing for now, as there have been a lot of pyinstaller revisions since. Though this may still be valid insight if the newest version runs into such issues again.
